### ISSUE TITLE
exp(p3): Social Influence as Intrinsic Motivation (Jaques et al.) — influence-shaped gradient

### ISSUE THREAD

**Issue by rjwalters** [rjwalters/bucket-brigade#290]
## Thesis context

The lit review identifies Jaques et al. (2019, ICML) "Social Influence as Intrinsic Motivation" as a directly-on-thesis algorithmic candidate. The paper adds an intrinsic reward proportional to how much agent i's action *causally influences* other agents' next actions. This explicitly encodes "your actions matter to others" into the gradient — a direct counter to defection's "your actions don't affect anyone but you" implicit assumption.

If our thesis says "the per-step advantage estimator misses the social structure," then giving each agent extra credit for *influencing* teammates is the surgical fix.

## Protocol

1. Implement influence-based intrinsic reward per Jaques et al. Each agent's reward = environmental reward + α × influence(self → others)
2. Sweep α ∈ {0, 0.1, 0.5, 1.0}; 3 seeds; 50 iters on `minimal_specialization`
3. Compare gap_closed against IPPO baseline + #283 potential-based shaping (the other "aligned-shaping" candidate)

## Verdict logic

| Outcome | Interpretation |
|---|---|
| Some α reaches gap_closed ≥ 0.5 | Influence-based shaping was missing. Thesis confirmed at "agents need to know their actions matter socially." |
| Flat across α | Either implementation issue or the misalignment is deeper than influence-shaped intrinsic |
| Inverted-U (small α helps, big α hurts) | Standard intrinsic-reward finding; document the sweet spot |

## Effort

~3-5 days. Influence computation requires counterfactual rollouts (what would agent j do if I'd taken a different action?) — adapt to our env.

## Sequencing

Phase 2. After cheaper experiments land.

## References

- Jaques, Lazaridou, Hughes, Gulcehre, Ortega, Strouse, Leibo, de Freitas (2019) "Social Influence as Intrinsic Motivation for Multi-Agent Deep Reinforcement Learning"
- `research_notebook/2026-05-17_thesis_misaligned_gradients.md`
- `research_notebook/2026-05-17_lit_review.md`

---

### Problem Statement

P3 thesis: IPPO converges to defection in bucket-brigade because each agent's per-step advantage signal does not encode that its action *matters socially*. We have empirically exhausted three "fix the local signal" levers (per-agent obs identity in PR #216, positional cost in PR #228, MAPPO centralized critic in PR #232) — PPO still closes ≤16% of the specialist-random gap on `minimal_specialization` / `positional_default`. The lit review (`research_notebook/2026-05-17_lit_review.md`) and thesis (`research_notebook/2026-05-17_thesis_misaligned_gradients.md`) identify Jaques et al. 2019 *Social Influence as Intrinsic Motivation* as a directly-on-thesis algorithmic candidate: it adds an intrinsic reward proportional to how much agent *i*'s action causally influences other agents' next-step action distributions. This bypasses the local advantage signal in exactly the direction the thesis predicts is missing.

### Paper citation

Jaques, N., Lazaridou, A., Hughes, E., Gulcehre, C., Ortega, P. A., Strouse, D. J., Leibo, J. Z., de Freitas, N. (2019). **Social Influence as Intrinsic Motivation for Multi-Agent Deep Reinforcement Learning**. *Proc. 36th ICML*, 3040–3049. arXiv:1810.08647.

### Math definition

For agent *i* taking action `a_i^t` at time *t*, with other agents *j ≠ i* taking joint action `a_{-i}^t`:

```
r_i^aug(t) = r_i^env(t) + alpha * influence_i(t)

influence_i(t) = sum_{j != i} KL( p(a_j^t | a_i^t, obs_j^t)  ||  p(a_j^t | obs_j^t) )

         = sum_{j != i} KL( p(a_j^t | a_i^t, obs_j^t)
                          ||  sum_{a'_i} p(a'_i | obs_i^t) * p(a_j^t | a'_i, obs_j^t) )
```

The marginal `p(a_j^t | obs_j^t)` is computed by marginalizing agent *j*'s conditional policy over agent *i*'s counterfactual actions weighted by `p(a'_i | obs_i^t)`. The paper's "Model of Other Agents" (MOA) head approximates `p(a_j^t | a_i^t, obs_j^t)` from rollout data so counterfactuals are cheap. The "basic influence" variant uses each agent's own policy network as the conditional model.

### Integration plan with current trainer

**Trainer**: `bucket_brigade/training/joint_trainer.py:145` (`JointPPOTrainer`).

Hook points (file:line refs to current `main`):

1. **Rollout buffer** — `JointPPOTrainer.collect_rollout` at `bucket_brigade/training/joint_trainer.py:393`. Add per-step storage of each agent's full action-distribution logits (already computed inside `_act_all` at `:354` — currently only the sampled action's log-prob is kept). New buffer field: `action_logits: Dict[int, Tensor[T, sum(action_dims)]]`.
2. **Influence reward computation** — new method `_compute_influence_reward(rollout) -> Dict[int, Tensor[T]]`. For each `(t, i, j)`:
   - Re-encode `obs_j^t` through agent *j*'s policy to get `p(a_j | a_i^t, obs_j^t)` (conditioning is implicit: `a_i^t` is part of agent *j*'s next-step obs via `last_actions`, see `bucket_brigade/envs/bucket_brigade_env.py:183`).
   - Compute the counterfactual marginal `p(a_j | obs_j^t)` by averaging over a Monte-Carlo set of `a'_i ~ p(a_i | obs_i^t)` (K=4 samples is the paper's default).
   - KL between the two categorical distributions, summed across the 3 action dims (`[house, mode, signal]`, `action_dims = [10, 2, 2]`, see `experiments/p3_specialization/train.py:92`).
3. **Augmented reward injection** — in `update()` at `:531`, before `_compute_advantages` at `:511`, replace `rollout.rewards[i]` with `rollout.rewards[i] + alpha * influence_i`. Advantages then propagate the influence credit through GAE normally.
4. **Config plumbing** — new `CellConfig` fields in `experiments/p3_specialization/train.py:57`:
   - `influence_coef: float = 0.0` (alpha; 0 preserves IPPO behavior bit-identically)
   - `influence_mc_samples: int = 4`
   - `influence_uses_moa: bool = False` (Phase 2 follow-on if basic influence shows lift)

**Counterfactual mechanism note**: bucket-brigade exposes other agents' previous-step actions to each agent via the `last_actions` field in the observation (see env `step()` at `bucket_brigade/envs/bucket_brigade_env.py:183`). This is what allows agent *j*'s next-step policy to be re-queried conditional on a counterfactual `a_i`. No env-side changes required — this is purely a trainer-side intervention.

### Hyperparameters to sweep

| Param | Values | Rationale |
|---|---|---|
| `influence_coef` (alpha) | {0.0, 0.1, 0.5, 1.0} | Paper sweep range; 0 is the IPPO control |
| `seed` | {0, 1, 2} | 3 seeds, matching #283 / #287 / #284 sibling cells |
| `num_iterations` | 50 | Matches the standard P3 50-iter × 2048-step protocol |
| `scenario` | `minimal_specialization` | The protocol-standard P3 scenario |

Total cells: 4 × 3 = 12. Estimated wall-clock on `COMPUTE_HOST_PRIMARY` (Mac Studio, CPU-bound): ~6 hours (each cell ~30 min).

### Success criterion

**Primary**: At least one `alpha > 0` cell achieves `gap_closed >= 0.5` on `minimal_specialization` (trailing-5 mean across seeds), where `gap_closed = (ppo_team_reward − random) / (specialist − random)` per `experiments/p3_specialization/analyze_231.py:127`. IPPO baseline currently sits at ≤0.16; ≥0.50 would be the first intervention to clear the "tier 1 succeeds" threshold defined at `analyze_231.py:173`.

**Secondary** (interpretive — does not gate verdict):
- Inverted-U over alpha (small helps, big hurts) → document the sweet spot.
- Per-agent action entropy (`entropy/agent_i` in `metrics.json`) stays > 0.5 — confirms influence reward did not collapse exploration.
- Pairwise action-KL (`tests/test_pairwise_action_kl.py` post-hoc) — should be similar to or greater than IPPO baseline; collapse would indicate influence reward is rewarding mimicry, not influence.

### Verdict logic (from original)

| Outcome | Interpretation |
|---|---|
| Some `alpha > 0` reaches `gap_closed >= 0.5` | Influence-shaped gradient was the missing ingredient. Thesis confirmed at "agents need to know their actions matter socially." |
| Flat across alpha (all cells within ±1 SE of IPPO baseline) | Either implementation issue or misalignment is deeper than influence-shaped intrinsic. Move to LOLA (#287). |
| Inverted-U (small alpha helps, big alpha hurts) | Standard intrinsic-reward finding; document sweet spot, file a follow-on for refined-alpha sweep. |

### Related issues

- **Sibling thesis interventions** (also `loom:curating` from the lit-review batch): #283 (potential-based shaping, Ng-Harada-Russell), #284 (COMA counterfactual baseline), #287 (LOLA), #288 (PBT diversity), #289 (HCA), #285 (BC+high-λ), #282 (high-λ GAE), #291 (single-agent control), #292 (toy 2x2 reproduction).
- **Upstream protocol context**: #193 (proposal: env-side reward shaping, currently `loom:blocked`), #234 (proposal: game v2 rules redesign, queued behind #193).
- **Diagnostic predecessors that exhausted the local-signal lever**: #145 (Phase 1 PPO-doesn't-learn), #174 (3-lever ablation), #183 (long-horizon), #197/#198/#199 (env-side fixes verified working — but IPPO still plateaus), #231 (MAPPO underperforms IPPO).

### Affected Files

- `bucket_brigade/training/joint_trainer.py` — add `influence_coef`, `influence_mc_samples` constructor args; extend `collect_rollout` to capture action logits; add `_compute_influence_reward`; inject augmented reward in `update()` before `_compute_advantages`.
- `experiments/p3_specialization/train.py` — extend `CellConfig` with `influence_coef`, `influence_mc_samples`, `influence_uses_moa`; pipe to `JointPPOTrainer`.
- `experiments/p3_specialization/run_sweep.py` — add `--influence-coef` sweep axis.
- `experiments/p3_specialization/analyze_290.py` (new) — load 12 cells, compute per-alpha bootstrap CIs on `gap_closed`, dump verdict JSON. Pattern after `analyze_270.py`.
- `tests/test_joint_trainer.py` — add unit test that with `influence_coef=0` the trainer is bit-identical to current behavior; second test that influence reward is non-negative and finite for a random rollout.
- `research_notebook/2026-05-18_issue290_social_influence_verdict.md` (new) — verdict writeup pattern matching `2026-05-17_issue272_bc_fit_only_verdict.md`.

### Test Plan

- [ ] Manual verification (single cell): `uv run python experiments/p3_specialization/train.py --scenario minimal_specialization --influence-coef 0.5 --seed 0 --num-iterations 5` — verify a 5-iter smoke run completes, `metrics.json` records `influence_reward/mean` non-zero, total reward = env reward + alpha * influence within numerical tolerance.
- [ ] Automated tests:
  - `tests/test_joint_trainer.py::test_influence_coef_zero_bit_identical` — with `influence_coef=0.0`, all per-iter metrics match current trainer to 1e-6.
  - `tests/test_joint_trainer.py::test_influence_reward_shape_and_finite` — for a 2048-step rollout, `_compute_influence_reward()` returns `Dict[int, Tensor[2048]]` with no NaN/Inf.
  - `tests/test_joint_trainer.py::test_influence_reward_zero_when_other_policy_constant` — if other agents' policies are forced to output a uniform distribution over actions, influence(self→others) is 0 (degenerate-conditional sanity check).
- [ ] Edge cases:
  - `influence_coef` very large (e.g. 100) — verify training doesn't NaN; clip if needed.
  - Single-agent scenario (would be #291's setup) — verify influence reward is exactly 0 (no other agents to influence) and falls back to plain IPPO.
- [ ] Full sweep verification (remote, `COMPUTE_HOST_PRIMARY`): run all 12 cells per "Hyperparameters to sweep"; verify `analyze_290.py` produces per-alpha `gap_closed` CIs; verdict written to research_notebook/.

### Compute resource note

All sweep cells must run on `COMPUTE_HOST_PRIMARY` per CLAUDE.md compute guidelines — IPPO P3 training is CPU-bound; 12 cells × 50 iters × 2048 steps is too large for localhost. Use the standard tmux+`run_sweep.py` pattern from `experiments/p3_specialization/run_issue262_sweep.sh` as a template.

### Effort

3–5 days end-to-end:
- 1 day: trainer plumbing + unit tests
- 1 day: sweep harness + analyzer
- 1 day: 6-hour remote sweep + monitoring
- 1 day: analysis writeup + verdict notebook
- 1 day buffer for debugging counterfactual computation (the trickiest piece)

### Sequencing

Phase 2. Behind cheaper sibling experiments where possible — but is independent of #287 (LOLA) and #284 (COMA) and can run in parallel.

**Comment by rjwalters** [comment]
Curating this issue per user request

**Comment by rjwalters** [comment]
Curation complete. Added paper citation (Jaques 2019 ICML arXiv:1810.08647), formal math definition of the influence intrinsic reward, integration plan with file:line refs to `JointPPOTrainer` (`bucket_brigade/training/joint_trainer.py:145, :393, :531`) and `CellConfig` (`experiments/p3_specialization/train.py:57`), hyperparameter sweep (alpha × seed = 12 cells, ~6 hours on COMPUTE_HOST_PRIMARY), success criterion (gap_closed ≥ 0.5 per `analyze_231.py:127`), Affected Files list (trainer + sweep harness + analyzer + tests + notebook), and a Test Plan with three concrete unit tests including the `influence_coef=0` bit-identity check. Linked to sibling lit-review issues (#283, #284, #287, #288, #289, #285, #282, #291, #292) and diagnostic predecessors (#145, #174, #183, #197-#199, #231). Dup checker flagged #291 and #284 as sibling exp(p3) issues from the same lit-review batch — not duplicates, they are distinct interventions. Awaiting human `loom:issue` approval.
